reconstruction/rich/src/main/java/org/jlab/rec/rich/Quaternion.java

```diff
@@ -0,0 +1,197 @@
+package org.jlab.rec.rich;
+
+import org.jlab.detector.geant4.v2.SVT.Matrix;
+
+import eu.mihosoft.vrl.v3d.Vector3d;
+
+//Class created by gangel
+//for the JLAB package 
+//
+// The quaternion here is defined as: 
+//
+public class Quaternion {
+	//Definition of the components for the Quaternion
+	public double x;
+	public double y;
+	public double z;
+	public double w;
+
+	// Temporary Quaternion   
+	private static Quaternion quatTp1 = new Quaternion();
+	private static Quaternion quatTp2 = new Quaternion();
+	private static Quaternion quatTp3 = new Quaternion();
+	private static Quaternion quatTp4 = new Quaternion();
+
+	public Quaternion() {
+	}
+
+	public Quaternion(double angle, Vector3d rotationAxis) {
+		x = rotationAxis.x * Math.sin(angle / 2);
+		y = rotationAxis.y * Math.sin(angle / 2);
+		z = rotationAxis.z * Math.sin(angle / 2);
+		w = Math.cos(angle / 2);
+	}
+
+	public Quaternion(double w, double x, double y, double z) {
+		this.x = x;
+		this.y = y;
+		this.z = z;
+		this.w = w;
+	}
+	/**
+	 *  Set the rotation Quaternion
+	 * @param angle the angle of rotation
+	 * @param rotationAxis the rotation axis (normal two the vector plane)
+	 */
+	public void set(double angle, Vector3d rotationAxis) {
+		x = rotationAxis.x * Math.sin(angle / 2);
+		y = rotationAxis.y * Math.sin(angle / 2);
+		z = rotationAxis.z * Math.sin(angle / 2);
+		w = Math.cos(angle / 2);
+	}
+
+	public void set (double w, double x, double y, double z) {
+		this.x = x;
+		this.y = y;
+		this.z = z;
+		this.w = w;
+	}
+
+	/**
+	 * Set the Quaternion using another one 
+	 * @param q
+	 */
+	public void set (Quaternion q) {
+		this.x = q.x;
+		this.y = q.y;
+		this.z = q.z;
+		this.w = q.w;
+	}
+
+	public double getSize() {
+		return Math.sqrt(w * w + x * x + y * y + z * z);
+	}
+
+	public void normalize() {
+		double sizeInv = 1 / getSize();
+		x *= sizeInv;
+		y *= sizeInv;
+		z *= sizeInv;
+		w *= sizeInv;
+	}
+	/** 
+	 * Multiplication between two Quaternions
+	 * @param qb the one that multiply
+	 * @param result of the multiplication 
+	 */
+	public void multiply(Quaternion qb, Quaternion result) {
```

Comment:
Looks like this isn't threadsafe?

Reply:
I think this is fine. It uses the this reference to access local data.